code/controllers/subsystem/runechat.dm

```diff
@@ -1,234 +1,16 @@
-				next.prev = prev
-			else
-				prev?.next = null
-				next?.prev = null
-			prev = next = null
-		scheduled_destruction = new_sched_destruction
-
-	// Ensure the scheduled destruction time is properly bound to avoid missing a scheduled event
-	scheduled_destruction = max(CEILING(scheduled_destruction, world.tick_lag), world.time + world.tick_lag)
-
-	// Handle insertion into the secondary queue if the required time is outside our tracked amounts
-	if (scheduled_destruction >= BUCKET_LIMIT)
-		BINARY_INSERT(src, SSrunechat.second_queue, datum/chatmessage, scheduled_destruction)
-		return
-
-	// Get bucket position and a local reference to the datum var, it's faster to access this way
-	var/bucket_pos = BUCKET_POS(scheduled_destruction)
-
-	// Get the bucket head for that bucket, increment the bucket count
-	var/datum/chatmessage/bucket_head = bucket_list[bucket_pos]
-	SSrunechat.bucket_count++
-
-	// If there is no existing head of this bucket, we can set this message to be that head
-	if (!bucket_head)
-		bucket_list[bucket_pos] = src
-		return
-
-	// Otherwise it's a simple insertion into the circularly doubly-linked list
-	if (!bucket_head.prev)
-		bucket_head.prev = bucket_head
-	next = bucket_head
-	prev = bucket_head.prev
-	next.prev = src
-	prev.next = src
-
-
-/**
-  * Removes this chatmessage datum from the runechat subsystem
-  */
-/datum/chatmessage/proc/leave_subsystem()
-	// Attempt to find the bucket that contains this chat message
-	var/bucket_pos = BUCKET_POS(scheduled_destruction)
-
-	// Get local references to the subsystem's vars, faster than accessing on the datum
-	var/list/bucket_list = SSrunechat.bucket_list
-	var/list/second_queue = SSrunechat.second_queue
-
-	// Attempt to get the head of the bucket
-	var/datum/chatmessage/bucket_head
-	if (bucket_pos > 0)
-		bucket_head = bucket_list[bucket_pos]
-
-	// Decrement the number of messages in buckets if the message is
-	// the head of the bucket, or has a SD less than BUCKET_LIMIT implying it fits
-	// into an existing bucket, or is otherwise not present in the secondary queue
-	if(bucket_head == src)
-		bucket_list[bucket_pos] = next
-		SSrunechat.bucket_count--
-	else if(scheduled_destruction < BUCKET_LIMIT)
-		SSrunechat.bucket_count--
-	else
-		var/l = length(second_queue)
-		second_queue -= src
-		if(l == length(second_queue))
-			SSrunechat.bucket_count--
-
-	// Remove the message from the bucket, ensuring to maintain
-	// the integrity of the bucket's list if relevant
-	if(prev != next)
-		prev.next = next
-		next.prev = prev
-	else
-		prev?.next = null
-		next?.prev = null
-	prev = next = null
+	var/list/datum/callback/message_queue = list()
 
-#undef BUCKET_LEN
-#undef BUCKET_POS
-#undef BUCKET_LIMIT
+/datum/controller/subsystem/timer/runechat/fire(resumed)
+	. = ..() //poggers
+	while(message_queue.len)
+		var/datum/callback/queued_message = message_queue[message_queue.len]
+		queued_message.Invoke()
+		message_queue.len--
+		if(MC_TICK_CHECK)
```

Comment:
Concerned about .length() checks here, or are you leaving them as .len on purpose for possible perf gains?

Reply:
len and length aren't quite the same IIRC